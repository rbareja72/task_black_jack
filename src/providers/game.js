import React, { useState } from 'react';
import { prepareDeck, multiply, shuffle } from '../library';

const useGame = () => {
  const [deck, setDeck] = useState([]);
  const [userTotal, setUserTotal] = useState(0);
  const [dealerTotal, setDealerTotal] = useState(0);
  const [userHand, setUserHand] = useState([]);
  const [dealerHand, setDealerHand] = useState([]);
  const [hasUserWon, setHasUserWon] = useState(null);
  const [userGameTotal, setUserGameTotal] = useState(0);
  const [dealerGameTotal, setDealerGameTotal] = useState(0);
  const getRandomCard = () => {
    const randomIndex = Math.random() * deck.length;
    const newDeck = [...deck];
    const card = newDeck.splice(randomIndex, 1);
    setDeck(newDeck);
    return card[0];
  };
  const drawAtRandom = (user) => {
    if (user === 'user') {
      setUserHand([
        ...userHand,
        getRandomCard(),
      ]);
    } else {
      setDealerHand([
        ...dealerHand,
        getRandomCard(),
      ]);
    }
  };
  const resetGame = () => {
    setDeck(shuffle(multiply(prepareDeck(), 3)));
    setDealerTotal(0);
    setUserTotal(0);
    setUserHand([]);
    setDealerHand([]);
    setHasUserWon(null);
    setDealerGameTotal(0);
    setUserGameTotal(0);
  };
  const reveal = () => {
    const userScore = calculateScore(userHand);
    let dealerScore = 0;
    const dealerCards = [];
    const newDeck = [...deck];
    while (dealerScore <= 21 && dealerScore <= userScore) {
      const randomIndex = Math.random() * newDeck.length;
      const card = newDeck.splice(randomIndex, 1);
      dealerCards.push(card[0]);
      dealerScore = calculateScore(dealerCards);
    }
    setDeck(newDeck);
    setDealerHand(dealerCards);
    setUserTotal(userScore);
    setDealerTotal(dealerScore);
    setHasUserWon((() => {
      if (userScore <= 21 && dealerScore >= 21) {
        return true;
      } else if (userScore < 21 && (21 - dealerScore) > (21 - userScore)) {
        return true;
      } else {
        return false;
      }
    })());
  };
  const calculateScore = (hand) => {
    let score = 0;
    const aces = hand.filter(card => card.display === 'A');
    const others = hand.filter(card => card.display !== 'A');
    others.forEach(card => { score += card.value; });
    aces.forEach(card => {
      if (score < 21 && score + card.value < 21) {
        score += card.value;
      } else {
        score += 1;
      }
    });
    return score;
  };
  const clearHandsAndUpdateScore = () => {
    setDealerTotal(0);
    setUserTotal(0);
    setUserHand([]);
    setDealerHand([]);
    if (hasUserWon) {
      setUserGameTotal(userGameTotal + 1);
    } else {
      setDealerGameTotal(dealerGameTotal + 1);
    }
    setHasUserWon(null);
  };
  return {
    deck,
    userHand,
    userTotal,
    hasUserWon,
    dealerHand,
    dealerTotal,
    userGameTotal,
    dealerGameTotal,
    reveal,
    resetGame,
    drawAtRandom,
    clearHandsAndUpdateScore,
  };
};

export const Context = React.createContext();

export const Provider = (props) => {
  const value = useGame();
  return (
    <Context.Provider value={value} {...props} />
  );
};
