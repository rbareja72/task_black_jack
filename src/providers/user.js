import React from 'react';

export const Context = React.createContext();

export const Provider = (props) => {
    const [user, setUser] = React.useState();

    return (
        <Context.Provider value={{ user, setUser }} {...props} />
    );
};
