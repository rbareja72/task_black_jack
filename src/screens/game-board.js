import React, { useEffect } from 'react';
import { View, Text, StyleSheet, Modal, FlatList } from 'react-native';
import { UserProvider, GameProvider } from '../providers';
import { Button } from '../components/button';
import { en } from '../assets/locale';
import colors from './../assets/colors';

import scaling from '../config/device/normalize';
const {
  normalize,
  widthScale,
  heightScale,
  moderateScale,
} = scaling;

export const GameBoard = () => {
  const [showModal, setShowModal] = React.useState(false);
  const user = React.useContext(UserProvider.Context);
  const game = React.useContext(GameProvider.Context);
  const {
    deck,
    userHand,
    drawAtRandom,
    reveal,
    hasUserWon,
    dealerHand,
    userTotal,
    dealerTotal,
    resetGame,
    clearHandsAndUpdateScore,
    userGameTotal,
    dealerGameTotal,
  } = game;
  useEffect(() => {
    if (hasUserWon || hasUserWon === false) {
      setShowModal(true);
    }
  }, [hasUserWon]);
  const onDraw = () => drawAtRandom('user');
  const onReveal = () => { reveal(); };
  const onRequestClose = () => setShowModal(false);
  const renderListItem = ({ item }) => (
    <View style={styles.listItem}>
      <Text style={styles.paragraph}>
        {item.title}
      </Text>
    </View>
  );
  const onNewPress = () => {
    onRequestClose();
    resetGame();
  };
  const onContinuePress = () => {
    onRequestClose();
    clearHandsAndUpdateScore();
  };
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{en.WELCOME + ', ' + user.user}</Text>
      <Text style={styles.subTitle}>{en.TOTAL_CARDS + ': ' + deck.length}</Text>
      <Text style={styles.subTitle}>{en.YOUR_HAND + ': ' + userHand.length}</Text>
      <Text style={styles.paragraph}>{en.RULES}</Text>
      {
        userGameTotal || dealerGameTotal
          ?
          <>
            <Text style={styles.subTitle}>
              {en.WON + ': ' + userGameTotal}
            </Text>
            <Text style={styles.subTitle}>
              {en.LOST + ': ' + dealerGameTotal}
            </Text>
          </>
          : null
      }
      <Button
        testID={'draw'}
        isDisabled={userHand.length > 20 || deck.length < 21}
        title={en.DRAW}
        onPress={onDraw}
        buttonStyle={styles.drawButton}
      />
      <Button
        testID={'reveal'}
        isDisabled={!userHand.length}
        title={en.REVEAL}
        onPress={onReveal}
        buttonStyle={styles.drawButton}
      />
      <Modal testID={'modal'} transparent visible={showModal} onRequestClose={onRequestClose} animated animationType={'slide'}>
        <View style={styles.modalContainer}>
          <Text style={styles.title}>{hasUserWon ? en.YOU_WIN : en.YOU_LOSE}</Text>
          <Text style={styles.subTitle}>{en.YOUR_HAND + ': ' + userTotal}</Text>
          <FlatList
            testID={'userHand'}
            style={styles.listStyle}
            data={userHand}
            keyExtractor={(item, index) => item.title + index}
            renderItem={renderListItem}
            bounces={false}
          />
          <Text style={styles.subTitle}>{en.DEALER_HAND + ': ' + dealerTotal}</Text>
          <FlatList
            testID={'dealerHand'}
            style={styles.listStyle}
            data={dealerHand}
            keyExtractor={(item, index) => item.title + index}
            renderItem={renderListItem}
            bounces={false}
          />
          <View style={styles.row}>
            <Button
              title={en.NEW_GAME}
              onPress={onNewPress}
              buttonContainer={styles.modalButton}
            />
            <Button
              isDisabled={deck.length < 21}
              title={en.CONTINUE}
              onPress={onContinuePress}
              buttonContainer={styles.modalButton}
            />
          </View>
        </View>
      </Modal>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    padding: moderateScale(8),
  },
  title: {
    fontSize: normalize(24),
    fontWeight: 'bold',
    paddingVertical: heightScale(20),
  },
  subTitle: {
    fontSize: normalize(18),
    paddingVertical: heightScale(12),
  },
  textInput: {
    height: heightScale(24),
    padding: 0,
    marginVertical: heightScale(24),
    borderBottomWidth: heightScale(1.5),
    borderColor: colors.blue,
    width: '80%',
    marginHorizontal: widthScale(12),
  },
  drawButton: {
    marginTop: heightScale(20),
  },
  paragraph: {
    fontSize: normalize(14),
  },
  modalContainer: {
    padding: moderateScale(16),
    margin: moderateScale(16),
    marginTop: heightScale(26),
    backgroundColor: colors.white,
    borderColor: colors.grey,
    borderWidth: moderateScale(1),
  },
  listStyle: {
    height: heightScale(150),
    borderWidth: moderateScale(1),
    borderColor: colors.grey,
  },
  listItem: {
    borderBottomColor: colors.grey,
    borderBottomWidth: heightScale(1),
    padding: heightScale(8),
  },
  row: {
    flexDirection: 'row',
    width: '100%',
    justifyContent: 'space-between',
    marginTop: heightScale(8),
  },
  modalButton: {
    width: '49%',
  },
});
