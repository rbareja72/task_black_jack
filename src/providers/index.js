import * as GameProvider from './game';
import * as UserProvider from './user';

export {
    GameProvider,
    UserProvider,
};
