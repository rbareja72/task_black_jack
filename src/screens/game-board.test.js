import 'react-native';
import React from 'react';
import { GameBoard } from './game-board';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { GameProvider, UserProvider } from '../providers';

describe('GameBoard', () => {

  const wrapper = shallow(
    <UserProvider.Provider>
      <GameProvider.Provider>
        <GameBoard />
      </GameProvider.Provider>
    </UserProvider.Provider>
  );

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

});

