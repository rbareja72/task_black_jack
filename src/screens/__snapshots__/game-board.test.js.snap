// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "setUser": [Function],
      "user": undefined,
    }
  }
>
  <Provider>
    <GameBoard />
  </Provider>
</ContextProvider>
`;
